host/traceh.c

```diff
@@ -67,88 +75,159 @@ void initialize_log_config()
         // inititalize if not already
         _log_level = _env2log_level();
         _log_file_name = getenv("OE_LOG_DEVICE");
+        _custom_log_format = getenv("OE_LOG_FORMAT");
+        _log_all_streams = getenv("OE_LOG_ALL_STREAMS");
+        _log_escape = getenv("OE_LOG_ESCAPE");
+        if (_custom_log_format)
+        {
+            // check that custom log format string terminates with a line return
+            size_t len = strlen(_custom_log_format);
+            if (_custom_log_format[len - 1] != '\n')
+            {
+                fprintf(
+                    stderr,
+                    "%s\n",
+                    "[ERROR] Custom log format does not end with a newline");
+                exit(1);
+            }
+        }
         _initialized = true;
     }
 }
 
-static void _write_header_info_to_stream(FILE* stream)
+static bool _escape_characters(
+    const char* log_msg,
+    char* log_msg_escaped,
+    size_t msg_size,
+    size_t max_msg_size)
 {
-    time_t t = time(NULL);
-    struct tm* lt = localtime(&t);
-
-    fprintf(
-        stream, "================= New logging session =================\n");
-    fprintf(stream, "%s", asctime(lt));
-    fprintf(
-        stream,
-        "https://github.com/openenclave/openenclave branch:%s\n",
-        OE_REPO_BRANCH_NAME);
-    fprintf(stream, "Last commit:%s\n\n", OE_REPO_LAST_COMMIT);
+    size_t idx = 0;
+    for (size_t i = 0; i < msg_size; i++)
+    {
+        int c = log_msg[i];
+        if (log_msg[i] == '\"')
+        {
+            // single quotes are OK for JSON
+            log_msg_escaped[idx] = (char)putchar('\'');
+        }
+        else if (log_msg[i] == '\\')
+        {
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            log_msg_escaped[idx] = '\\';
+        }
+        else if (!isprint(c))
+        {
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            switch (log_msg[i])
+            {
+                case '\a':
+                    log_msg_escaped[idx] = 'a';
```

Comment:
\a is not listed in the JSON grammar here: https://json.org/
Can you double-check the list of characters to be escaped?

Reply:
Thanks, I believe I escape only what is needed now.